python/sbp/client/handler.py

```diff
@@ -14,151 +14,44 @@
-
-  Parameters
-  ----------
-  read : port
-    Stream of bytes to read from.
-  write : port
-    Stream of bytes to write to.
-  """
-  def __init__(self, read, write, verbose=False):
-    self.read = read
-    self.write = write
-    self.verbose = verbose
-
-  def readall(self, size):
-    """
-    Read until all bytes are collected.
-
-    Parameters
-    ----------
-    size : int
-      Number of bytes to read.
-    """
-    data = ""
-    while len(data) < size:
-      data += self.read(size - len(data))
-    return data
-
-  def receive(self):
-    """
-    Read and build SBP message.
-    """
-    # preamble - not readall(1) to allow breaking before messages, empty input
-    preamble = self.read(1)
-    if not preamble:
-      return None
-    elif ord(preamble) != SBP_PREAMBLE:
-      if self.verbose:
-        print "Host Side Unhandled byte: 0x%02x" % ord(preamble)
-      return None
-    # hdr
-    hdr = self.readall(5)
-    msg_crc = crc16(hdr)
-    msg_type, sender, msg_len = struct.unpack("<HHB", hdr)
-    # data
-    data = self.readall(msg_len)
-    msg_crc = crc16(data, msg_crc)
-    # crc
-    crc = self.readall(2)
-    crc, = struct.unpack("<H", crc)
-    if crc != msg_crc:
-      print "crc mismatch: 0x%04X 0x%04X" % (msg_crc, crc)
-      return None
-    return SBP(msg_type, sender, msg_len, data, crc)
-
-  def send(self, msg_type, data, sender):
-    """
-    Build and write SBP message.
-
-    Parameters
-    ----------
-    msg_type : int
-      SBP message type.
-    data : string
-      SBP message payload.
-    sender : int
-      SBP sender id.
-    """
-    msg = struct.pack("<BHHB", SBP_PREAMBLE, msg_type, sender, len(data))
-    msg += data
-    crc = crc16(msg[1:])
-    msg += struct.pack("<H", crc)
-    self.write(msg)
-
-class ReceiveThread(threading.Thread):
-  """
-  ReceiveThread
-
-  The :class:`ReceiveThread` class provides a processing thread for
-  framing and handling SBP messages.
-
-  Parameters
-  ----------
-  receive : fn
-    Receiving function returns SBP messages.
-  call : fn
-    Callback functions.
-  """
-  def __init__(self, receive, call):
-    super(ReceiveThread, self).__init__()
-    self.receive = receive
-    self.call = call
-    self.daemon = True
-    self.stopping = False
-
-  def stop(self):
-    """
-    Stop the thread before the next message.
-    """
-    self.stopping = True
-
-  def run(self):
-    """
-    Reading and handling loop.
-    """
-    while not self.stopping:
-      try:
-        msg = self.receive()
-        if msg is not None:
-          if msg.msg_type:
-            self.call(msg)
-      except SystemExit:
-        break
+import weakref
+from Queue import Queue
 
 class Handler(object):
   """
   Handler
 
   The :class:`Handler` class provides an interface for connecting handlers
-  to a driver providing SBP messages.
+  to a driver providing SBP messages.  Also provides queued and filtered
+  iterators for synchronous, blocking use in other threads.
```

Comment:
Can we also have a asynchronous, nonblocking handling of messages? If the handler throws them on the queue, having an async callback that returns immediately would be useful in a few cases.